Sources/NIOHPACK/HuffmanCoding.swift

```diff
@@ -0,0 +1,221 @@
+            }
+            
+            return state.offset
+        }
+    }
+    
+    private mutating func appendSym_fast(_ sym: HuffmanTableEntry, _ state: inout _EncoderState) {
+        // will it fit as-is?
+        if sym.nbits == state.remainingBits {
+            state.buffer[state.offset] |= UInt8(sym.bits)
+            state.offset += 1
+            state.remainingBits = state.offset == state.buffer.count ? 0 : 8
+        } else if sym.nbits < state.remainingBits {
+            let diff = state.remainingBits - sym.nbits
+            state.buffer[state.offset] |= UInt8(sym.bits << diff)
+            state.remainingBits -= sym.nbits
+        } else {
+            var (code, nbits) = sym
+            
+            nbits -= state.remainingBits
+            state.buffer[state.offset] |= UInt8(code >> nbits)
+            state.offset += 1
+            
+            if nbits & 0x7 != 0 {
+                // align code to MSB
+                code <<= 8 - (nbits & 0x7)
+            }
+            
+            // we can short-circuit if less than 8 bits are remaining
+            if nbits < 8 {
+                state.buffer[state.offset] = UInt8(truncatingIfNeeded: code)
+                state.remainingBits = 8 - nbits
+                return
+            }
+            
+            // longer path for larger amounts
+            if nbits > 24 {
+                state.buffer[state.offset] = UInt8(truncatingIfNeeded: code >> 24)
+                nbits -= 8
+                state.offset += 1
+            }
+            
+            if nbits > 16 {
+                state.buffer[state.offset] = UInt8(truncatingIfNeeded: code >> 16)
+                nbits -= 8
+                state.offset += 1
+            }
+            
+            if nbits > 8 {
+                state.buffer[state.offset] = UInt8(truncatingIfNeeded: code >> 8)
+                nbits -= 8
+                state.offset += 1
+            }
+            
+            if nbits == 8 {
+                state.buffer[state.offset] = UInt8(truncatingIfNeeded: code)
+                state.offset += 1
+                state.remainingBits = state.offset == state.buffer.count ? 0 : 8
+            } else {
+                state.remainingBits = 8 - nbits
+                state.buffer[state.offset] = UInt8(truncatingIfNeeded: code)
+            }
+        }
+    }
+    
+    private mutating func ensureBitsAvailable(_ bits: Int, _ buffer: inout ByteBuffer) {
+        let bytesNeeded = bits / 8
+        if bytesNeeded <= buffer.writableBytes {
+            // just zero the requested number of bytes before we start OR-ing in our values
+            buffer.withUnsafeMutableWritableBytes { ptr in
+                ptr.baseAddress!.assumingMemoryBound(to: UInt8.self).assign(repeating: 0, count: bytesNeeded)
+            }
+            return
+        }
+        
+        // one extra byte to ensure we have space to fill at the end of the encoded string, if we need it
```

Comment:
This comment doesn't appear to correspond to any actual code. Is that an oversight?